Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py

```diff
@@ -0,0 +1,714 @@
+                errors[irow, icol, :] = self.ws.readE(ispec)
+        # get x bin centers
+        xvals = self.ws.readX(int(ispecs[irow_peak, icol_peak]))
+        if len(xvals) > self.ws.blocksize():
+            xvals = 0.5 * (xvals[:-1] + xvals[1:])  # convert to bin centers
+        return xvals, signal, errors, irow_peak, icol_peak, detector_edges, detids
+
+
+def is_peak_mask_valid(peak_mask, peak_label, npk_min, density_min, nrow_max, ncol_max,
+                       min_npixels_per_vacancy, max_nvacancies):
+    if peak_label == 0:
+        return False, PEAK_MASK_STATUS.NO_PEAK
+    if peak_mask.sum() < npk_min:
+        return False, PEAK_MASK_STATUS.NPIX_MIN
+    ncol = np.sum(peak_mask.sum(axis=0) > 0)
+    if ncol > ncol_max:
+        return False, PEAK_MASK_STATUS.NCOL_MAX
+    nrow = np.sum(peak_mask.sum(axis=1) > 0)
+    if nrow > nrow_max:
+        return False, PEAK_MASK_STATUS.NROW_MAX
+    density = peak_mask.sum() / (ncol * nrow)
+    if density < density_min:
+        return False, PEAK_MASK_STATUS.DENSITY_MIN
+    if does_peak_have_vacancies(peak_mask, min_npixels_per_vacancy, max_nvacancies):
+        return False, PEAK_MASK_STATUS.VACANCY_MAX
+    return True, PEAK_MASK_STATUS.VALID
+
+
+def does_peak_have_vacancies(peak_mask, min_npixels_per_vacancy=2, max_nvacancies=1):
+    labeled_array, num_features = label(~peak_mask)
+    nvac = 0
+    for ifeature in range(1, num_features + 1):
+        feature_mask = labeled_array == ifeature
+        if not does_mask_touch_window_edge(feature_mask):
+            # is a vacancy within the peak region
+            if feature_mask.sum() >= min_npixels_per_vacancy:
+                nvac += 1
+                if nvac > max_nvacancies:
+                    return True
+    return False
+
+
+def does_mask_touch_window_edge(peak_mask):
+    return any(np.sum(peak_mask, axis=0).astype(bool)[[0, -1]]) or any(np.sum(peak_mask, axis=1).astype(bool)[[0, -1]])
+
+
+def get_nearest_non_masked_index(mask, irow, icol):
+    r, c = np.nonzero(mask)
+    dist = ((r - irow) ** 2 + (c - icol) ** 2)
+    imins = np.where(dist == dist.min())
+    return np.array(r[imins]), np.array(c[imins])
+
+
+def find_bg_pts_seed_skew(signal, ibg_seed=None):
+    if ibg_seed is None:
+        ibg_seed = np.arange(signal.size)
+    # sort and grow seed
+    isort = np.argsort(-signal[ibg_seed])  # descending order
+    istart = 0
+    iend = len(signal)
+    prev_skew = moment(signal[ibg_seed[isort[istart:iend]]], 3)
+    for istart in range(1, iend):
+        this_skew = moment(signal[ibg_seed[isort[istart:iend]]], 3)
+        if this_skew >= prev_skew:
+            istart -= 1
+            break
+        else:
+            prev_skew = this_skew
+    return ibg_seed[isort[istart:iend]], ibg_seed[isort[:istart]]  # bg, non-bg
+
+
+def calc_snr(signal, error_sq):
+    return np.sum(signal) / np.sqrt(np.sum(error_sq))
+
+
+def optimise_window_intens_over_sig(signal, error_sq, ilo, ihi, ntol=8):
```

Comment:
is the window here a tof window as opposed to a detector window? If yes then would help clarify if you add "tof" in the name

Reply:
Yep, I will make the function name something like
`optimise_tof_window_intens_over_sig`
as you suggest!